[NNPA] Memory reduction of stickified constant by stickifying at file writing

imaihal opened the pull request:
This PR reduces memory usage for NNPA compilation. Current main branch creates stickified data in ZHighConstPropagationPass and keeps the data until compilation finish. This PR sets original data, not stickified data, in ZHighConstPropagationPass. Then, in the KrnlToLLVMPass, stickfied data is created and stored in the file, and deleted after writing into the file.  

<img width="412" alt="image" src="https://github.com/user-attachments/assets/d4f93155-a3aa-4a46-b469-f24938eccd8c">

imaihal commented:
@jenkins-droid test this please.

AlexandreEichenberger commented on src/Accelerators/NNPA/Dialect/ZHigh/ZHigh.td:
```diff
@@ -862,11 +862,16 @@ def ZHighStickifiedConstantOp:ZHigh_Op<"StickifiedConstant", [Pure]> {
     This operator produces a constant tensor to store stickified data.
+    `value` attribute has original constant or stickified constant.
+    `stickified` attribute indicates the `value` is already stickified or not.
+    `allzero` attribute indicates the `value` will be all zero or not.
```
Quick question" if it is `allzero` do we need to actually have the data, or it would be sufficient to have the shape as there is no need to store 10,000 times the value zero?

imaihal replied:
Originally(in main branch) the value zeros are created and set in value attribute in ZHighStickifiedConstantOp. This consumes a lot of memory in some models. So, I created `allzero` attribute and set nullptr in value attribute in ZHighStickifiedConstantOp . Then, when generating LLVMIR, the value zeros are created and stored soon in the file. 
To answer your question, currently the value zeros are stored in the file. This PR delays the creation of the value zeros.

tungld replied:
If it is zero, you can use a splat dense attribute for `value` that stores one zero only and we don't need a new attribute `allzero`.

imaihal replied:
I removed `allzero` attribute. Thanks!

AlexandreEichenberger reviewed (commented):
High level question: do we need both the old and the new method? Are there any drawbacks to  the new method? If so, can you add this in the PR so that it is documented and we know why we have both? Ideally, we could drop the older approach.

imaihal commented:
@AlexandreEichenberger 
> High level question: do we need both the old and the new method? Are there any drawbacks to the new method? If so, can you add this in the PR so that it is documented and we know why we have both? Ideally, we could drop the older approach.

I checked inference time and accuracy was no problem in some LLMs. Also, I confirmed the results for model zoo models using RunONNXModelZoo.py are the same with main branch both on CPU and NNPA. So I will drop the older approach. Thanks.

tungld commented on src/Accelerators/NNPA/Transform/ZHigh/ZHighConstPropagation.cpp:
```diff
@@ -91,44 +95,54 @@ ZHighStickifiedConstantOp createConstantForStick(PatternRewriter &rewriter,
+        /*allzero=*/nullptr,
+        /*alignment=*/rewriter.getI64IntegerAttr(4096));
+  } else {
+    ArrayRef<int64_t> shape =
```
Stickify code here is the same as the code in `getBuffer()` for `ZLowStickifiedConstant`. Could you create a share function for it? Thanks!

tungld commented on src/Dialect/Krnl/KrnlOps.cpp:
```diff
@@ -812,6 +814,50 @@ MutableOperandRange KrnlSpecializedKernel::getLoopRefs() {
   return getLoopsMutable();
 }
 
+ArrayRef<char> getRawData(KrnlGlobalOp &op) {
```
Again, we have some duplicated code here. To avoid that, this function can be a function in `ConstantOpInterface` and we can name it `getValueBuffer()`, and `getBuffer()` can be renamed to `getOutputBuffer()`.

tungld reviewed (commented):
Thank you for the PR! This is my first batch of comments.

tungld commented on src/Accelerators/NNPA/Compiler/NNPACompilerOptions.cpp:
```diff
@@ -100,4 +100,10 @@ llvm::cl::opt<bool> nnpaEnableSaturation("nnpa-saturation",
+    "nnpa-delay-stickified-const-generation",
+    llvm::cl::desc(
+        "Generate stickified constants when storing constants to file. This "
+        "will be able to reduce peak memory usage. Default is true."),
```
I see you always check `if (storeConstantsToFile && nnpaDelayStickifiedConstGen)`. So, it seems this option only works when `store-constant-to-file` is enabled. I recommend to turn it off at the beginning of the compilation if `store-constant-to-file` is off, then you need to check `if (nnpaDelayStickifiedConstGen)` only.

By that way you don't need to check

imaihal replied:
I removed this option. Thanks.

imaihal commented on src/Accelerators/NNPA/Conversion/ZHighToZLow/ZHighToZLow.cpp:
```diff
@@ -686,7 +692,7 @@ struct ZHighToZLowUnstickOpLowering : public ConversionPattern {
 
 //===----------------------------------------------------------------------===//
-// Lower ZHigh Stickified Constant to KrnlGlobal
+// Lower ZHigh Stickified Constant to ZLow Stickified Constant
```
I removed the flag. Now all Zhigh stickified constant is lowered to Zlow stickified constant. Thanks!

imaihal commented on src/Accelerators/NNPA/Conversion/ZHighToZLow/ZHighToZLow.cpp:
```diff
@@ -713,36 +719,60 @@ struct ZHighToZLowStickifiedConstantOpLowering : public ConversionPattern {
+              /*allzero=*/zhighStickifiedConstOp.getAllzeroAttr(),
+              /*alignment=*/zhighStickifiedConstOp.getAlignmentAttr());
+      result = constantOp.getResult();
+    } else {
```
That is because I kept original implementation, but now I removed the option. Now all ZHighStickifiedConst is lowered to ZLowStickifiedConst. Thanks!

imaihal commented on src/Accelerators/NNPA/Dialect/ZLow/ZLowOps.cpp:
```diff
@@ -358,6 +366,152 @@ void ZLowBatchNormOp::getEffects(
+    zlowStickifiedConstantOp.removeValueAttr();
+    zlowStickifiedConstantOp.removeStickifiedAttr();
+  } else if (auto allZero = zlowStickifiedConstantOp.getAllzeroAttr()) {
+    if (allZero) {
```
I removed `allzero` attribute. Now this code was removed. Thanks!

imaihal commented on src/Accelerators/NNPA/Dialect/ZLow/ZLow.td:
```diff
@@ -547,4 +548,21 @@ def ZLowConvertF32ToDLF16VectorOp:ZLow_Op<"vec_f32_to_dlf16", [Pure]> {
+                       OptionalAttr<BoolAttr>:$stickified,
+                       OptionalAttr<BoolAttr>:$allzero,
+                       DefaultValuedAttr<I64Attr, "4096">:$alignment);
+  let results = (outs ZMemRef:$output);
```
@tungld I thought the operands can be included in `ConstantOpInterface.td` using `let arguments`, but it seems we can't use it. Do you know another way to realize it?

tungld replied:
I see. It allows to define functions only. Thanks for the confirmation!

chentong319 commented:
First high-level question: source of the problem for memory surge.
Here are what happened in the transformation:
 1. In the original implementation, when a onnx.Constant is stickified, the constant folding transformation will create a krnl.GlobalOp to hold the stickified value. 
 2. The memory usage may be increased by two possible reasons: 1) both the onnx.Constant and krnl.GlobalOp are alive, 2) the krnl.GlobalOp is significantly larger than onnx.Constant due to the layout. I think the reason1) is the major factor. After onnx to zhigh and then to krnl, I think we can free the pool for onnx.Constant (That's the purpose of using disposable pool for onnx.Constant. 
 3. To confirm: the problem is that the space for stickified constant is doubled, one for onnx.constant and one for krnl.global during the process from onnx to zhigh until all onnx ops are lowered to krnl.
 4. The root of the problem is that space for onnx.Constant cannot be released one-by-one. If a onnx.Constant is only used by stickification, we should free the onnx.Constant right after the krnl.Global is created.

imaihal commented on src/Accelerators/NNPA/Dialect/ZLow/ZLowOps.cpp:
```diff
@@ -358,6 +366,152 @@ void ZLowBatchNormOp::getEffects(
+          })
+      .Default(
+          [&](Attribute attr) { llvm_unreachable("Unsupported data type."); });
+}
```
Done. Thanks!

imaihal commented on src/Accelerators/NNPA/Dialect/ZLow/ZLowOps.cpp:
```diff
@@ -358,6 +366,152 @@ void ZLowBatchNormOp::getEffects(
+}
+
+/// MLIR type to zDNN type.
+zdnn_data_types mlirTypeToZDNNType(Type elementType) {
```
Done. Thanks!

chentong319 commented:
Question about the Solution:
1. The stickification is delayed to lowering constant to llvm.  In this lowering, can the space for krnl.Global be freed one-by-one while new llvm constant is created? I am comparing the situation when onnx.Constant is converted to stickified krnl.Global in onnx to zhigh in original code.
2. Following the previous point, I'd like to know whether the benefit of this PR comes from storing constant in file instead of in-memory, or from delaying the stickification operation.

chentong319 commented:
Comment on the implementation:
1. To my understanding, the ConstantOpinterface introduced by this PR is not really needed. You may follow the approach how we convert similar op, for example operations in RNN group. Sometime, it is better to localize the implementation. Another possible implementation is to use inheritance in Op definition. We didn't use it for ONNX because the definition is derived from onnx package, which does not specify the inheritance. ZLow.stickifiedConstant is a subclass of Krnl.Global with extra stickification description.
2. I am wondering whether we can use the ElementAttr directly for krnl.Global from onnx.Constant. To my understanding, the difference of krnl.Global and onnx.Constant is the type, memref vs. tense. But the value representation for constant is the same. Can we avoid copying the data? Also the dump out of IR will be uniformed.

imaihal commented:
@chentong319 Thanks for the confirmation.

>     1. In the original implementation, when a onnx.Constant is stickified, the constant folding transformation will create a krnl.GlobalOp to hold the stickified value.

Yes.

>     2. The memory usage may be increased by two possible reasons: (1) both the onnx.Constant and krnl.GlobalOp are alive, (2) the krnl.GlobalOp is significantly larger than onnx.Constant due to the layout. I think the reason (1) is the major factor. After onnx to zhigh and then to krnl, all onnx.ConstantOps are converted to krnl.Global, and then  the pool for onnx.ConstantOps can be freed (That's the purpose of using disposable pool for onnx.Constant. )

Yes. The major factor should be (1). 

>     3. To confirm: the problem is that the space for stickified constant is doubled, one for onnx.constant and one for krnl.global during the process from onnx to zhigh until all onnx ops are lowered to krnl. Did you see a plateau in memory size during that period of transformation?

In original implementation, even after all onnx ops are lowered to krnl, the memory size is doubled. I think onnx.ConstantOps are not released. Once stickified data is created, the memory size keep double until the end of execution. So, in this PR, stickified data is created just before writing to file, then deleted it just after writing. This means memory size is not reduced by this PR if not writing to file.

>     4. The root of the problem is that space for onnx.Constant cannot be released one-by-one. If a onnx.Constant is only used by stickification, we should free the onnx.Constant right after the krnl.Global is created.

I think so, but from my testing, even after deleting any operation, memory size was not reduced now.

imaihal commented:
> Question about the Solution:
>     1. The stickification is delayed to lowering constant to llvm.  In this lowering, can the space for krnl.Global be freed one-by-one while new llvm constant is created? I am comparing the situation when onnx.Constant is converted to stickified krnl.Global in onnx to zhigh in original code.
>     2. Following the previous point, I'd like to know whether the benefit of this PR comes from storing constant in file instead of in-memory, or from delaying the stickification operation.

I updated the figure in this PR since previous figure might be confusing. 
The solution does not use krnl.Global. Instead, zlow.StickifiedConstant is used and lowered to llvm.  When krnl.Global is used, memory usage can not be reduced. KrnlGlobal.cpp is also modified in this PR, but this modification is for using the ConstantOpInterface. 
When zlow.StickifiedConstant is lowered to llvm, the data in value attribute is stickified, stored into a file, then deleted after writing. Original non-stickified data is alive, but stickified data is not alive. 
The benefit comes from delaying the stickification, but it only benefits when storing into file.

chentong319 commented:
The previous figure is fine. I knew that this PR replaced some krnl.GlobalOps (result of stickifiedConstant) with zlowstickifiedConstant. The previous figure showed the reason why the ConstantOpInterface is introduced.

imaihal commented on test/mlir/accelerators/nnpa/driver/ccfd.mlir:
```diff
@@ -5,15 +5,13 @@
+CHECK-DAG: zlow.stickifiedConstant
+CHECK-DAG: zlow.stickifiedConstant
 CHECK-DAG: memref.alloc
-CHECK-NEXT: zlow.stick
```
Thanks for comments. I fixed it. I think I mistakenly  removed it.

imaihal commented:
@chentong319 Thanks for the comments. I'm now looking into RNN group and inheritance in Op definition.

chentong319 commented:
Our RNN group did not use inheritance. We just use a common utility function for core part of different RNN operations.

imaihal commented:
I'm considering to use common functions. 

I'm investigating how we can commonize `extractConstantsToFile()` for KrnlGlobalOp and ZLowStickifiedConstantOp. Currently it is called [`here`](https://github.com/onnx/onnx-mlir/blob/8cce1cb99ecccbb517933f05a0757a0fb69e072a/src/Conversion/KrnlToLLVM/ConvertKrnlToLLVM.cpp#L864). In the function, `ConstantOpInterface` is used for KrnlGlobalOp and ZLowStickifiedConstOp. I'm considering to commonize it using template, then call it here such as `extractConstantsToFile<KrnlGlobalOp>` or `extractConstantsToFile<ZLowStickifiedConstantOp>`, but I think we can't set operation name here.  Do you know good way?

chentong319 commented:
If you use the Op matching rule for rewrite, the op is cast automatically.
In your this implementation,  you are traverse the ops by yourself.  The op has to be cast. 
Then you can call the member functions and pass the results as parameters to the common utility function. 
If you feel it is too complicated, it is okay to me to use ConstantOpInterface (perhaps with another less general name).

imaihal commented:
@chentong319 Thanks for the investigations and comments! 

I tried to implement some other way such as common functions and Trait, but I couldn't write it clean. So, I would like to keep using OpInterface. 
How about using `KrnlGlobalOpInterface`? I used the opInterface to use the same functions with KrnlGlobalOp in zlowStickifiedConstantOp. I can replace if you have better candidate.

imaihal commented on src/Dialect/Krnl/KrnlOps.cpp:
```diff
@@ -812,6 +814,50 @@ MutableOperandRange KrnlSpecializedKernel::getLoopRefs() {
+
+ArrayRef<char> KrnlGlobalOp::getBuffer() {
+  ArrayRef<char> rawData;
+  if (auto krnlGlobalOp = mlir::dyn_cast<KrnlGlobalOp>(getOperation())) {
```
Yes. So, updated this function. Thanks!

chentong319 commented on src/Accelerators/NNPA/Conversion/ZHighToZLow/ZHighToZLow.cpp:
```diff
@@ -713,36 +697,27 @@ struct ZHighToZLowStickifiedConstantOpLowering : public ConversionPattern {
-           "The stickified tensor's buffer size and MemRef's size mismatched");
-
-    // Create a KrnlGlobalOp.
-    KrnlGlobalOp constantGlobal =
```
Keep the previous implementation with KrnlGlobalOp in comment or `if false` branch, if you do not want to create an option to control the choice.

imaihal replied:
OK. Is this because we may reuse the previous implementation in the future?

imaihal replied:
I created directive `NNPA_ZHIGH_STICKIFIEDCONST_GEN` to keep the original implementation. Currently commented out, but I confirmed it works when enabling this code.

chentong319 replied:
yes, we may reuse or compare the original implementation.

imaihal commented on src/Accelerators/NNPA/Dialect/ZLow/ZLowOps.cpp:
```diff
@@ -358,6 +366,152 @@ void ZLowBatchNormOp::getEffects(
+  MLIRContext *context = getOperation()->getContext();
+  PatternRewriter rewriter(context);
+  ZLowStickifiedConstantOp zlowStickifiedConstantOp =
+      mlir::cast<ZLowStickifiedConstantOp>(getOperation());
```
Done. Thanks

imaihal commented on src/Accelerators/NNPA/Dialect/ZLow/ZLowOps.cpp:
```diff
@@ -358,6 +366,152 @@ void ZLowBatchNormOp::getEffects(
+      mlir::cast<ZLowStickifiedConstantOp>(getOperation());
+  ArrayRef<char> ret;
+  if (zlowStickifiedConstantOp.getValueAttr() &&
+      zlowStickifiedConstantOp.getStickifiedAttr()) {
```
I changed 'stickified' attribute as mandatory attribute. So, now only check getValueAttr() here. Thanks!

imaihal commented on src/Interface/ConstantOpInterface.hpp:
```diff
@@ -0,0 +1,28 @@
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
```
Remove the pragma. Thanks.

imaihal commented on src/Interface/ConstantOpInterface.td:
```diff
@@ -0,0 +1,78 @@
+          "void", "updateBuffer", (ins )
+      >,
+      InterfaceMethod<"Get the constant value from value attribute. ",
+          "::mlir::ArrayRef<char>", "getBuffer", (ins )
```
Updated descriptions in the OpInterface. Thanks!

imaihal commented on src/Interface/ConstantOpInterface.td:
```diff
@@ -0,0 +1,78 @@
+
+  let methods = [
+      InterfaceMethod<"Update the constant value to value attribute.",
+          "void", "updateBuffer", (ins )
```
Changed the function name to `updateValueAttr()`. and updated the description. 
Since the stickified data is created when the data is stored in the file, if it is small and not stored in the file, the data in the value attribute is not stickified data. This function is used to set stickified data to value attribute.

imaihal commented:
Could you review again?

AlexandreEichenberger commented:
@chentong319 Do you mind reviewing it? You have been very involved with this PR. Thanks

chentong319 reviewed (approved):
LGTM!

chentong319 reviewed (approved):
LGTM!

chentong319 commented:
Something is wrong with my git account. I can not approve the PR.
